apollo-router/src/introspection.rs

```diff
@@ -1,142 +1,163 @@
+            ",
             )
-            .into(),
-        })?;
-
-        let response = Response::builder().data(introspection_result).build();
-
-        self.cache.insert(query, response.clone()).await;
-
-        Ok(response)
+            .extension_code("MIXED_INTROSPECTION")
+            .build();
+        graphql::Response::builder().error(error).build()
     }
-}
-
-#[cfg(test)]
-mod introspection_tests {
-    use std::sync::Arc;
-
-    use router_bridge::planner::IncrementalDeliverySupport;
-    use router_bridge::planner::QueryPlannerConfig;
-
-    use super::*;
 
-    #[tokio::test]
-    async fn test_plan_cache() {
-        let query_to_test = r#"{
-            __schema {
-              types {
-                name
-              }
+    async fn cached_intropsection(
+        &self,
+        schema: &Arc<spec::Schema>,
+        key: &QueryKey,
+        doc: &ParsedDocument,
+    ) -> graphql::Response {
+        let storage = match self {
+            IntrospectionCache::Enabled { storage } => storage,
+            IntrospectionCache::Disabled => {
+                let error = graphql::Error::builder()
+                    .message(String::from("introspection has been disabled"))
+                    .extension_code("INTROSPECTION_DISABLED")
+                    .build();
+                return graphql::Response::builder().error(error).build();
             }
-          }"#;
-        let schema = include_str!("../tests/fixtures/supergraph.graphql");
-        let expected_data = Response::builder().data(42).build();
+        };
+        let query = key.filtered_query.clone();
+        // TODO:  when adding support for variables in introspection queries,
+        // variable values should become part of the cache key.
+        // https://github.com/apollographql/router/issues/3831
+        let cache_key = query;
+        if let Some(response) = storage.get(&cache_key, |_| unreachable!()).await {
+            return response;
+        }
+        let schema = schema.clone();
+        let doc = doc.clone();
+        let response =
+            tokio::task::spawn_blocking(move || Self::execute_intropsection(&schema, &doc))
+                .await
+                .expect("Introspection panicked");
+        storage.insert(cache_key, response.clone()).await;
+        response
+    }
 
-        let planner = Arc::new(
-            Planner::new(
-                schema.to_string(),
-                QueryPlannerConfig {
-                    incremental_delivery: Some(IncrementalDeliverySupport {
-                        enable_defer: Some(true),
-                    }),
-                    graphql_validation: true,
-                    reuse_query_fragments: Some(false),
-                    generate_query_fragments: None,
-                    debug: None,
-                    type_conditioned_fetching: false,
-                },
+    fn execute_intropsection(schema: &spec::Schema, doc: &ParsedDocument) -> graphql::Response {
```

Comment:
```suggestion
    fn execute_introspection(schema: &spec::Schema, doc: &ParsedDocument) -> graphql::Response {
```

Reply:
I seem to like that particular typo, I made it in two places